dev-lang/rust/rust_bin-1.21.0.recipe

```diff
@@ -0,0 +1,94 @@
+SUMMARY="Modern and safe systems programming language"
+DESCRIPTION="Rust is a systems programming language that runs blazingly fast, prevents almost all crashes*, and eliminates data races."
+HOMEPAGE="https://www.rust-lang.org/"
+COPYRIGHT="2017, The Rust Project Developers"
+LICENSE="MIT"
+REVISION="1"
+
+case "$effectiveTargetArchitecture" in
+	x86)
+SOURCE_URI="http://dl.rust-on-haiku.com/dist/$portVersion/rust-$portVersion-i686-unknown-haiku.tar.xz"
+CHECKSUM_SHA256="465db013876ead01686c991f1fa63ef4a6687d3cabbcde81f5a756e367989cdf"
+SOURCE_DIR="rust-1.21.0-i686-unknown-haiku"
+		;;
+	x86_64)
+SOURCE_URI="http://dl.rust-on-haiku.com/dist/$portVersion/rust-$portVersion-x86_64-unknown-haiku.tar.xz"
+CHECKSUM_SHA256="cbb7f705398b76c5e2dd18119cdb9af2c042720449479b482a39309fb270ded8"
+SOURCE_DIR="rust-$portVersion-x86_64-unknown-haiku"
+		;;
+	*)
+SOURCE_URI="http://dl.rust-on-haiku.com/dist/$portVersion/rustc-$portVersion-src.tar.xz"
+CHECKSUM_SHA256="7741c2f3652813b7ecd0233305a5ca0de561fea2849049651a79521d2c437f0f"
+SOURCE_DIR="rustc-$portVersion-src"
+		;;
+esac
+
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
+SECONDARY_ARCHITECTURES="x86"
+DISABLE_SOURCE_PACKAGE=yes
+
+cargoVersion="0.22.0"
+rlsVersion="0.121.0"
+
+PROVIDES="
+	rust$secondaryArchSuffix = $portVersion
+	cmd:rustc = $portVersion
+	cmd:rustdoc = $portVersion
+	cmd:rust_gdb = $portVersion
+	cmd:rust_lldb = $portVersion
+	cmd:cargo$secondaryArchSuffix = $cargoVersion
```

Comment:
It turns out that adding `commandBinDir` and `commandSuffix` works fine here.
Shall we use it?
And shall we tweak the `PROVIDES` to have rust_bin?
The changes for the above suggestions are:
```diff
 DISABLE_SOURCE_PACKAGE=yes
 
+commandSuffix=$secondaryArchSuffix
+commandBinDir=$binDir
+if [ "$targetArchitecture" = _x86_gcc2 ]; then
+	commandSuffix=
+	commandBinDir=$prefix/bin
+fi
+
 cargoVersion="0.22.0"
 rlsVersion="0.121.0"
 
 PROVIDES="
-	rust$secondaryArchSuffix = $portVersion
+	rust_bin$secondaryArchSuffix = $portVersion
-	cmd:rustc = $portVersion
-	cmd:rustdoc = $portVersion
-	cmd:rust_gdb = $portVersion
-	cmd:rust_lldb = $portVersion
+	cmd:rustc$commandSuffix = $portVersion
+	cmd:rustdoc$commandSuffix = $portVersion
+	cmd:rust_gdb$commandSuffix = $portVersion
+	cmd:rust_lldb$commandSuffix = $portVersion
-	cmd:cargo$secondaryArchSuffix = $cargoVersion
+	cmd:cargo$commandSuffix = $cargoVersion
-	cmd:rls = $rlsVersion
+	cmd:rls$commandSuffix = $rlsVersion
 	"
```
and later, in `INSTALL`:
```diff
 	./$module/install.sh                               \
 			--prefix=$prefix                           \
+			--bindir=$commandBinDir                    \
 			--docdir=$developDocDir                    \
```
```diff
-	# move the cargo and binaries (in case of a secondary arch)
-	if [ -n "$secondaryArchSuffix" ]; then
- 		mkdir -p $binDir
- 		mv $prefix/bin/cargo $binDir/cargo
- 	fi
```

Reply:
I don't quite understand what you are doing here, but whatever it is, it should comply with the following:
1. cargo depends on being able to call an executable called `rustc`. 
2. `rustc` will determine its sysroot relative to its own location. Specifically it looks for the `rustlib` directory at the `../lib/rustlib` location. That's why on x86_gcc2h only cargo has this suffix, not rustc. 

As per 2, the rationale that cargo has the suffix, is to remind the user to perform a setarch x86 before trying to compile rust projects, as not to cause problems with using the wrong compiler and linker.

Reply:
OK, I got it :)
(I'll update my comment asap.)

Reply:
In case it is better, for a x86 secondary arch build, to have all the runtimes in `/system/bin/x86/` then we should:
* keep your version of the _"move the cargo and binaries"_ bloc;
* ignore my suggestion to add definitions for `commandBinDir` and `commandSuffix`;
* replace all occurences of `$commandSuffix` by `$secondaryArchSuffix`.

Reply:
Actually it is better not to have binaries in /system/bin/x86, but in /system/bin.

Reply:
Great! 
My suggestions [above](https://github.com/haikuports/haikuports/pull/2647#discussion_r199521705) handles this nicely :)

BTW, I forgot to answer your question about why I'm suggesting to define `commandBinDir` and `commandSuffix`. It's to make sure that, on x86_64-hybrid, the `cmd:*` provided by `rust_bin_x86` won't shadow the `cmd:*` provided by `rust_bin`.
This will make sure recipes which might have e.g. `cmd:rustc` in their `BUILD_REQUIRES` will pull `rust_bin` and not `rust_bin_x86`.
(On x86_gcc2 there is no risk of the `cmd:*` for the secondary arch build shadowing those of the primary arch... because rust is not available on x856_gcc2 primary arch.)

Yeah, as I wrote in #2690, nobody will try to build `rust_x86` on `x86_64-hybrid`, but using a `commandSuffix` will allow to have a valid recipe in all cases and avoid a situation where two different builds of rust_bin (that is, for x64_64 primary arch and for x86 secondary arch) would provide the same `cmd:*`.